cms/static/js/views/import.js

```diff
@@ -2,241 +2,319 @@
+         *
+         */
+        var initEventListeners = function () {
+            window.onbeforeunload = function () {
+                if (current.stage <= STAGE.UNPACKING) {
+                    return gettext('Your import is in progress; navigating away will abort it.');
+                }
+            }
+        };
 
         /**
-         * Manipulate the DOM to reflect current status of upload.
-         * @param {int} stageNo Current stage.
+         * Stores in a cookie the current import data
+         *
+         * @param {boolean} [completed=false] If the import has been completed or not
          */
-        var updateStage = function (stageNo){
-            var all = $('ol.status-progress').children();
-            var prevList = all.slice(0, stageNo);
-            _.map(prevList, function (elem){
-                $(elem).
-                    removeClass("is-not-started").
-                    removeClass("is-started").
-                    addClass("is-complete");
-                updateCog($(elem), false);
-            });
-            var curList = all.eq(stageNo);
-            curList.removeClass("is-not-started").addClass("is-started");
-            updateCog(curList, true);
+        var storeImport = function (completed) {
+            $.cookie(COOKIE_NAME, JSON.stringify({
+                file: file,
+                date: moment().valueOf(),
+                completed: completed || false
+            }));
         };
 
         /**
-         * Check for import status updates every `timeout` milliseconds, and update
-         * the page accordingly.
-         * @param {string} url Url to call for status updates.
-         * @param {int} timeout Number of milliseconds to wait in between ajax calls
-         *     for new updates.
-         * @param {int} stage Starting stage.
+         * Sets the Import on the "success" status
+         *
+         * If it wasn't already, marks the stored import as "completed",
+         * and updates its date timestamp
          */
-        var getStatus = function (url, timeout, stage) {
-            var currentStage = stage || 0;
-            if (currentStage > 1) { CourseImport.okayToNavigateAway = true; }
-            if (CourseImport.stopGetStatus) { return ;}
-
-            if (currentStage === 4) {
-                // Succeeded
-                CourseImport.displayFinishedImport();
-                $('.view-import .choose-file-button').html(gettext("Choose new file")).show();
-            } else if (currentStage < 0) {
-                // Failed
-                var errMsg = gettext("Error importing course");
-                var failedStage = Math.abs(currentStage);
-                CourseImport.stageError(failedStage, errMsg);
-                $('.view-import .choose-file-button').html(gettext("Choose new file")).show();
-            } else {
-                // In progress
-                updateStage(currentStage);
+        var success = function () {
```

Comment:
Yeah I noticed it too and I don't like too much either because it seems the object interface is lacking consistency, but there is a reason in terms of the interface exposing the methods that the client script can actually use:

* `start`, `resume`, `reset` are obviously methods used by the client to control the object
* `error` is a method that was supposed to be internal like `success` (as the object itself can invoke it internally when it detects something went wrong), but it's also necessary for the client to be able to signal the object that something went wrong on its end, and that's why it needs to be exposed

`success` on the other end is completely dependent on the internal logic of the object, as only `CourseImport` can tell if an import has been completed successfully or not. Even if I made it a public method so the script client could invoke it, it would have no use for it (at least from how the whole thing works now)

Anyway I just wanted to explain my reasoning behind why it is a private method, I'm curious what your thoughts on it are. You think it's fine to move it even if the public interface has a kind of a useless method? I'm fine about it either way :)

> This would also allow to make `deferred` an internal member of CourseImport.

I'm not sure I follow, isn't `deferred` already private?